### ISSUE TITLE
race detection as a PR check

### ISSUE THREAD

**Issue by andrewhsu** [moby/moby#39649]
Perhaps a test suite specifically to exercise race detection: https://blog.golang.org/race-detector

Would be good to setup a PR check for this, but if it takes a long time to run, then perhaps as a nightly job on master branch?

**Comment by andrewhsu** [comment]
To help in detecting issues like #39643

**Comment by AkihiroSuda** [comment]
I think this should be nightly or maybe weekly

**Comment by thaJeztah** [comment]
> I think this should be nightly or maybe weekly

Agreed, I don't think we currently want this to run on every PR, or on master; at least, not if it makes the build _fail_

I'm thinking along these lines;

- Add a new "nightly" job on Jenkins; if we can "detect" (from the jenkinsfile) that it's running as part of that nightly job, we can enable additional tests that we would normally skip on pull-requests (and/or master). Thinking about things, such as:
    - run tests with `-race` (the topic of this issue)
    - run stress/load tests (we don't have those currently, but we could write some)
- Define test-matrices for different combinations of the components;
    - test against `golang tip` to catch regressions early
    - add a stage to run on a non-debian based distro (actually, I think we should have such a stage in our regular job)
    - run the old `integration-cli` tests with different combinations of the CLI (18.03-ee, 18.09-ee, 19.03 (ce/ee), master); this would help catching possible regressions (we would have to make some changes to be able to fetch the EE Cli from somewhere 😅)
    - same as the above, but run the "e2e" tests from https://github.com/docker/cli. I think that's also where we should move/add (public) tests that are intended to test the whole end-to-end cycle
    - run the tests with different versions of containerd (1.1.x, 1.2.x, 1.3.x); containerd [promises a stable API](https://github.com/containerd/containerd/blob/v1.3.0/RELEASES.md#public-api-stability), so we should test that (ideally)
- If integrating all of those in a single Jenkinsfile, perhaps having a separate Jenkinsfile to work with would be better (adding a `nighty.Jenkinsfile`, and perhaps a `nightly.mk` Makefile to extend the existing one to keep things a bit separated)
- Run the Kubernetes e2e/validation/conformance suite (not too familiar with the procedure for those)

@andrewhsu wdyt?

**Comment by thaJeztah** [comment]
Actually, with some smart writing of that Jenkinsfile, we could _trigger_ it on each PR, but disable all stages, and leaving it up to maintainers to run (some) those stages if they see a need to run them on a specific PR. For example, in https://github.com/moby/moby/pull/40050 I made a change to force all validation steps to run on master, but adding a checkbox to also allow running them on a pull request (when clicking `rebuild` in Jenkins);

<img width="467" alt="Screenshot 2019-10-11 at 14 19 48" src="https://user-images.githubusercontent.com/1804568/66651116-88d9b600-ec32-11e9-8717-f354595b26b9.png">


Jenkins also has a `input` option for Jenkinsfiles, which allow interactively setting options before a test runs (with a configurable timeout to cancel); I _think_ that can also be used by a bot (which would allow selecting those options directly on GitHub through editing a comment); See https://medium.com/disney-streaming/jenkins-pipeline-with-dynamic-user-input-9f340fb8d9e2 for an example
